### ISSUE TITLE
Example for calculating param values from model dump

### ISSUE THREAD

**Issue by Mahdi-Hosseinali** [microsoft/LightGBM#2320]
The documentation has a nice guideline about [dealing with over fitting](https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html), however, it is not clear how to set the mentioned parameters. For example, user does not have feedback about `num_leaves`, `min_data_in_leaf`, `min_gain_to_split`, or `max_depth` of a booster that is using default parameters but overfits. Although all four parameters are easy to calculate from the booster object, it would be nice to have them as inbuilt functions rather than copying the functions in every problem.

I've written the python code and would be happy to contribute them if I know how to do it properly. I understand that it might be a better to implement them in the core so all APIs have access to them. Or it might be too much of a hassle to guide me on how to implement them rather than translating the core by a conversant contributor, so here's the code. It outputs all four parameters as lists for each tree, so the user can plot them and decide what would be a better value.

```python
def depth(d):
    if isinstance(d, dict):
        return 1 + (max(map(depth, d.values())) if d else 0)
    return 0

def min_gain_to_split(tree):
    if 'split_index' in tree:
        lst = [tree['split_gain'],
               min_gain_to_split(tree['left_child']),
               min_gain_to_split(tree['right_child']) ]
        return min([i for i in lst if i>0])
    return -1

def min_data_in_leaf(tree):
    if 'leaf_count' in tree:
        return tree['leaf_count']
    return min(
        min_data_in_leaf(tree['left_child']),
        min_data_in_leaf(tree['right_child']) )

def get_tree_inf(booster):
    trees = booster.dump_model()['tree_info']
    leaves, depths, min_data, min_gains = [], [], [], []
    for tree in trees:
        leaves.append(tree['num_leaves'])
        tree_struct = tree['tree_structure']
        depths.append(depth(tree_struct))
        min_data.append(min_data_in_leaf(tree_struct))
        min_gains.append(min_gain_to_split(tree_struct))
    return [leaves, depths, min_data, min_gains]
```

**Comment by StrikerRUS** [comment]
@Mahdi-Hosseinali Thanks for sharing this useful snippet! To be honest, I'm not sure this functionality should be a part of API. Maybe [`examples`](https://github.com/microsoft/LightGBM/tree/master/examples/python-guide)is better place for it?

**Comment by Mahdi-Hosseinali** [comment]
that sounds good, which example do you think it would fit better?

**Comment by StrikerRUS** [comment]
@Mahdi-Hosseinali I think either [`advanced example`](https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py) (there is already `dump_model()` call there, BTW) can be modified or separate new one can be created.

https://github.com/microsoft/LightGBM/blob/9cf6b8283d6257bf5b044197de7c77c709c85def/examples/python-guide/advanced_example.py#L70

**Comment by Mahdi-Hosseinali** [comment]
Both options sounds equally well. Since you are a collaborator, I trust that your decision is based on a better understanding of the repository. If you decide to go with a separate example and want me to put together one, I'd be happy to. The second option could be linked in the parameter tuning part of the documentation (link in the original post), for a faster learning pace of new users (such as myself).

**Comment by StrikerRUS** [comment]
@Mahdi-Hosseinali To tell you the truth, the best option I can imagine is a separate example about hyperparameters tuning, which will include both very basic parts (e.g. picking the best param value according to grid search) and some more advanced technics like your proposal. But it will require definitely more free time. If you are ready to take it, it'll be great! If not, the part of the existing advanced example is very good too.

**Comment by Mahdi-Hosseinali** [comment]
I'll be busy for next couple of weeks, but will be happy to do it then. Do you have any particular dataset in mind that'd be a good test case?

**Comment by StrikerRUS** [comment]
Sure, any time you can! I think any dataset from our [`examples` folder](https://github.com/microsoft/LightGBM/tree/master/examples) is OK. Or you can use scikit-learn built-in datasets.

**Comment by StrikerRUS** [comment]
@Mahdi-Hosseinali Any updates?

**Comment by Mahdi-Hosseinali** [comment]
> @Mahdi-Hosseinali Any updates?

Sorry for the long delay, I was hoping to get to it after my graduation, but took on a job and the company has a very tight deadline. However, it should be over by the end of November and I'll put this as my top priority. I hope that is okay with you and sorry for not sticking to my word.

**Comment by StrikerRUS** [comment]
@Mahdi-Hosseinali No problem! Thanks! I just wanted to know that this issue is not abandoned.

**Comment by StrikerRUS** [comment]
Closed via #2592.
